### PR TITLE
Add haskell-ide-engine.

### LINKED ISSUE #31697: Request for new package: haskell-ide-engine
## Issue description

I would like to have [`haskell-ide-engine` (HIE)](https://github.com/haskell/haskell-ide-engine) added to `nixpkgs`, but I fear I am not knowledgeable enough to add it myself.

The Haskell IDE Engine is described in its home page by the following:
> This project aims to be the universal interface to a growing number of Haskell tools, providing a full-featured and easy to query backend for editors and IDEs that require Haskell-specific functionality.

It is a requirement for Haskell development in [Visual Studio Code](https://code.visualstudio.com/) using the extension [Haskell Language Server](https://marketplace.visualstudio.com/items?itemName=alanz.vscode-hie-server), whose repository is [Haskell Language Server Client](https://github.com/alanz/vscode-hie-server).

### PR DESCRIPTION
**Pull request by shlevy**
Fixes #31697
References haskell/haskell-ide-engine#357

### CODE REVIEW

**Comment by shlevy** [comment]
@Fresheyeball @imalsogreg Can you test and see if this fits the bill of your bounty?

**Comment by Fresheyeball** [comment]
@shlevy While I love you forever for putting `hie` into `nixpkgs`. That was not the issue with the bounty. This is the issue with the bounty: https://github.com/haskell/haskell-ide-engine/issues/357

**Comment by shlevy** [comment]
@Fresheyeball Yes, but if you pop `haskell-ide-engine` from the same package set as your package into your nix-shell env I believe this should work.

**Comment by Fresheyeball** [comment]
@shlevy ok, let me test.

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix [hunk 1]
```diff
@@ -71,4 +71,12 @@ self: super: {
   # https://github.com/aristidb/aws/issues/238
   aws = doJailbreak super.aws;
 
+  # There's no 8.2 testsuite
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: The changes to `configuration-ghc-8.2.x.nix` seem to be orthogonal to the addition of the HIE package set. Can you please submit those fixes separately (or commit them directly) to take them out of this PR for easier reviewing?

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix [hunk 2]
```diff
@@ -71,4 +71,12 @@ self: super: {
...
+  ghc-exactprint = dontCheck super.ghc-exactprint;
+
+  # Missing FlexibleContexts in test suite
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: Is upstream aware of this issue? If they are, do you have a link to the upstream ticket so that we can see the current state of this bug?

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix [hunk 3]
```diff
@@ -71,4 +71,12 @@ self: super: {
...
+  monad-memo = dontCheck super.monad-memo;
+
+  haddock-api = super.haddock-api.override
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: Is this override specific to GHC 8.2.2? It looks like it might belong into `configuration-nix.nix` or `configuration-common.nix`.

File: pkgs/development/haskell-modules/initial-packages.nix [hunk 4]
```diff
@@ -0,0 +1,2 @@
+args@{ pkgs, stdenv, callPackage }: self:
+  (import ./hackage-packages.nix args self) // (import ./hie-packages.nix args self)
```
Comment (peti) [inline pkgs/development/haskell-modules/initial-packages.nix]: Please reverse the order of that set update to make sure that no entry in hie-packages can ever replace an entry of the same name in hackage-packages.

File: pkgs/development/haskell-modules/patches/haddock-library-system-attoparsec.patch [hunk 5]
```diff
@@ -0,0 +1,46 @@
+diff -Naur haddock-library-1.4.4-orig/haddock-library.cabal haddock-library-1.4.4/haddock-library.cabal
```
Comment (peti) [inline pkgs/development/haskell-modules/patches/haddock-library-system-attoparsec.patch]: Are those patch files maybe available for download from somewhere so that we don't have to check them into our repository?

File: pkgs/development/haskell-modules/patches/hie-build-plugin-haddock.patch [hunk 6]
```diff
@@ -0,0 +1,23 @@
+commit ae0fb7a79a8912cf7a40bee418db239a7188805c
```
Comment (peti) [inline pkgs/development/haskell-modules/patches/hie-build-plugin-haddock.patch]: Ditto. Can this file be downloaded from somewhere else?

File: pkgs/development/haskell-modules/patches/hie-testsuite.patch [hunk 7]
```diff
@@ -0,0 +1,40 @@
+diff --git a/test/HaRePluginSpec.hs b/test/HaRePluginSpec.hs
```
Comment (peti) [inline pkgs/development/haskell-modules/patches/hie-testsuite.patch]: Ditto. Can this file be downloaded from somewhere else?

**Review by peti — DISMISSED** [review]
Looks good to me. I'd be happy if we can avoid checking the patch files into our repo, though. Is that possible?

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix (hunk 1 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: 👍

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix (hunk 2 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: I'll open a ticket, thanks.

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix (hunk 3 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: Right, I confused myself about the order of overrides, I didn't want to overwrite the old versions uses in 7.10 and 7.8. Will fix.

File: pkgs/development/haskell-modules/hie-packages.nix [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+        sha256 = "0z7r3l4j5a1brz7zb2rgd985m58rs0ki2p59y1l9i46fcy8r9y4g";
+      };
+    cabal-helper = self.cabal-helper_hie;
+    haddock-library = self.haddock-library_1_4_4;
+    ghc-dump-tree = self.ghc-dump-tree_hie;
+    ghc-mod = self.ghc-mod_hie;
+    HaRe = self.HaRe_hie;
+in
+  { # This isn't really hie-specific, should probably go in cabal2nix
+    haddock-library_1_4_4 = callPackage
```
Comment (shlevy) [inline pkgs/development/haskell-modules/hie-packages.nix]: @peti What's the right way to add this to hackage-packages.nix?

File: pkgs/development/haskell-modules/initial-packages.nix (hunk 4 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/initial-packages.nix]: 👍

File: pkgs/development/haskell-modules/patches/haddock-library-system-attoparsec.patch (hunk 5 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/patches/haddock-library-system-attoparsec.patch]: This one isn't available elsewhere. I should probably open an issue upstream, this is only needed because install doesn't work right yet with sublibraries.

File: pkgs/development/haskell-modules/patches/hie-build-plugin-haddock.patch (hunk 6 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/patches/hie-build-plugin-haddock.patch]: This is from a PR, I'll fetch it

File: pkgs/development/haskell-modules/patches/hie-testsuite.patch (hunk 7 above)
Comment (shlevy) [reply pkgs/development/haskell-modules/patches/hie-testsuite.patch]: This is a workaround, not a real fix, and doesn't exist anywhere yet.

File: pkgs/development/haskell-modules/hie-packages.nix (hunk 8 above)
Comment (peti) [reply pkgs/development/haskell-modules/hie-packages.nix]: Just add it to the `extra-packages` section in `pkgs/development/haskell-modules/configuration-hackage2nix.yaml`. Then it will appear in the main package set within a couple of days (i.e. whenever I think of merging `haskell-updates` into `master`).

**Comment by shlevy** [comment]
@peti good now?

**Comment by shlevy** [comment]
This is now on top of #32993 to avoid the need to patch `haddock-library`

**Comment by domenkozar** [comment]
@shlevy are `hie-packages.nix` hand-crafted?

**Comment by shlevy** [comment]
@domenkozar Yes, though the bulk of it could be scripted pretty easily if updates become obnoxious

**Comment by vcunat** [comment]
Add to the next haskell-update iteration to avoid double re-building (and re-checking)?

**Comment by shlevy** [comment]
@vcunat I think that applies more to #32993 (which this PR is on top of) than this, but yes.

**Comment by vcunat** [comment]
Right, I didn't notice a commit was from a different PR.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 9]
```diff
@@ -1019,4 +1019,14 @@ self: super: {
       { patches = (drv.patches or []) ++ [ patch ];
         editedCabalFile = null;
       });
+
+  # https://github.com/haskell/cabal/issues/4969
+  haddock-library_1_4_4 = dontHaddock
+    (super.haddock-library_1_4_4.override
+       { # The 'attoparsec' haddock-library depends on is its own
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: This issue is fixed in 40a3d775a5887ccc944d009032f70f0dd2b891f7. `cabal2nix-1.7.1` no longer generates the bogus dependency.

**Review by peti — APPROVED** [review]
Great work! Please feel free to merge. I have added one more minor nit that you may want to address beforhand.

**Comment by michalrus** [comment]
Silly question: will this handle Cabal@Nix projects, too? :heart_eyes_cat:  I.e. the Nix (not NixOS) part in https://github.com/haskell/haskell-ide-engine/issues/357.

**Comment by shlevy** [comment]
@michalrus I don't really understand what you mean by `Cabal@Nix` projects, but the point of this PR was to address that issue 😄 . The idea is that your nix-shell (or nix-buffer, or whatever) env will include `haskell-ide-engine` from the same package set as your actual project, so you know it matches.

**Comment by michalrus** [comment]
Pure Cabal projects with deps managed by Nix. =]

Awesome, so it should just work. ♥

Thank you! ♥ ♥

Finally I can get rid of Intero and those hacks, which are ±useless, cf. https://github.com/michalrus/intero-nix-shim/issues/7.

**Comment by peti** [comment]
`haskell-ide-engine` does not compile because the following dependency fails: https://hydra.nixos.org/build/66889657

**Comment by imalsogreg** [comment]
@shlevy I'm seeing
```
...
building path(s) ‘/nix/store/gbi2cwqamac15insqhb8yzl338bwmzk0-haddock-api-2.18.1’, ‘/nix/store/gwq9pbhfq9z5349zyakb2bn7x5pcjf80-haddock-api-2.18.1-data’, ‘/nix/store/vcj4h08c41yasfdhxh4i92as1r8jmal0-haddock-api-2.18.1-doc’
setupCompilerEnvironmentPhase
Build with /nix/store/3b1axvsyr53f4f5nihvpxv3p4hskb3j7-ghc-8.2.2.
ln: failed to create symbolic link '/nix/store/gbi2cwqamac15insqhb8yzl338bwmzk0-haddock-api-2.18.1/lib/links/libHShaddock-library-1.4.4-BXedb9tEAJEKpUW5xu3aRK-attoparsec-ghc8.2.2.dylib': File exists
ln: failed to create symbolic link '/nix/store/gbi2cwqamac15insqhb8yzl338bwmzk0-haddock-api-2.18.1/lib/links/libHShaddock-library-1.4.4-G8iPfxLjJ1KJncaQAcQooZ-ghc8.2.2.dylib': File exists
builder for ‘/nix/store/zjq2bjsnvn6mg50izlnav0mkwjg6m93y-haddock-api-2.18.1.drv’ failed with exit code 1
building path(s) ‘/nix/store/9wcrbw3w1dgmy0g209jxs3adfykmi2fm-happy-1.19.8’, ‘/nix/store/x7adjlzcz42dhmzw60ixf9m7s84lwjvy-happy-1.19.8-doc’
cannot build derivation ‘/nix/store/ksb034msgmaq1pcbayij0vw4ny0gmdhl-hie-haddock-0.1.0.0.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/8jw9zx2l5id8lfgcg563w54mv8dvgzk0-haskell-ide-engine-0.1.0.0.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/8jw9zx2l5id8lfgcg563w54mv8dvgzk0-haskell-ide-engine-0.1.0.0.drv’ failed
```
(`nix-env -f ~/nixpkgs -iA haskellPackages.haskell-ide-engine -j1`) on macos.

**Comment by shlevy** [comment]
@peti This works as of 89e882f956bcddf01ea2d69a569156bf93dcb9c6 . Testing on latest master now.

**Comment by shlevy** [comment]
@Ericson2314 any idea about the issue @imalsogreg is seeing? Seems to have something to do with the dynamic loading stuff on darwin.

**Comment by shlevy** [comment]
@peti Ah, this is due to the lts bump. Will take a look soon.

**Comment by Ericson2314** [comment]
@shelvy I think that's a Darwin impurity from previous runs.

**Comment by shlevy** [comment]
@Ericson2314 ? What impure dir is being referenced there?

**Comment by Ericson2314** [comment]
@shlevy this sounds to be true, but I swear this hit this when working on the darwin dynamic linking stuff. The previous attempt earlier in the build (say `preConfigure`) would but a a bunch of symlinks in there to avoid using as many RPATHs`. If the first  build failed then on the second attempt it would fail because for some reason $out wasn't cleaned out and the symlinks would still be there.

**Comment by shlevy** [comment]
@Ericson2314 That would be a huge bug in nix if true...

**Comment by shlevy** [comment]
@imalsogreg I think it should be fixed in https://github.com/NixOS/nixpkgs/pull/33333

**Comment by shlevy** [comment]
(but there's still the hoogle issue to fix 😄 )

**Comment by imalsogreg** [comment]
@shlevy @Ericson2314 Looking at the error message from cabal trying to solve version constraints, I tried (a) patching hoogle 5.0.15 to accept haskell-src-exts_1_19_1 and (b) dropping the nixpkgs override that sets haskell-src-exts to 1.20.1.

This worked on my ubuntu machine. But the patch to hoogle introduces some ugly CPP to allow the newest hoogle to support both haskell-src-exts versions. So I'm wondering, is there another way to support hie-hoogle here?

**Comment by shlevy** [comment]
@imalsogreg I'll fix that side of things in a later PR.

**Comment by imalsogreg** [comment]
@shlevy making sure I understand: I should _not_ try to get the `hoogle` patch (which loosens its haskell-src-exts deps to include 1.19.1) merged?

**Comment by shlevy** [comment]
Oh, if you can get it merged upstream that'd also be good.

**Comment by shlevy** [comment]
@imalsogreg should be fixed in https://github.com/NixOS/nixpkgs/commit/707cfbf39a27058c14de5af36fb3b22febb7935b

**Comment by peti** [comment]
@shlevy, according to https://hydra.nixos.org/job/nixpkgs/haskell-updates/haskellPackages.haskell-ide-engine.x86_64-linux/all, the package hasn't seen a successful build yet, apparently because the test suite assumes network access. Would you mind taking a look at that issue?

**Comment by shlevy** [comment]
f12f2ed44c66ac65b0307f3562152e9a45251d11

**Comment by michalrus** [comment]
Hmm, I can’t build it on current unstable, 5402412b97247bcc0d2b693e159d55d114d1327b:

```
Warning:
    This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
      package ghc-mod-core (ghc-mod-core-5.9.0.0) requires haskell-src-exts-1.19.1-GkJUFo8Rp3b1KlAdoTXU6c
      package hlint (hlint-2.0.15-DnfSXCOnlvhJh25E4Cb1Bg) requires haskell-src-exts-1.20.1-835K5nW7Qg0K3DUFrUYhiW
      package haskell-src-exts-util (haskell-src-exts-util-0.2.1.2-4UhAxVuyQajDfxD9gJGXOC) requires haskell-src-exts-1.20.1-835K5nW7Qg0K3DUFrUYhiW

*** abort because of serious configure-time warning from Cabal
builder for ‘/nix/store/bkry8lqkj605av0ybmi1qismb42h5j4w-ghc-mod-core-5.9.0.0.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/mk2i2g7lgg4f7lwpf0m1jbr2idv282f8-haskell-ide-engine-0.1.0.0.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/mk2i2g7lgg4f7lwpf0m1jbr2idv282f8-haskell-ide-engine-0.1.0.0.drv’ failed
```

**Comment by michalrus** [comment]
@shlevy ^

**Comment by shlevy** [comment]
I'll work on this but I'm not sure when I'll have time.
